File: modules/loc8-mesh/android/src/main/java/expo/modules/loc8mesh/MeshDiagnostics.kt
// MeshDiagnostics.kt
// Loc8Mesh
//
// Bounded, privacy-preserving evidence recorder for the preregistered MESH-01
// three-phone field experiment. Observation only: no routing or wire behavior.

package expo.modules.loc8mesh

import android.os.SystemClock
import java.security.MessageDigest

internal class MeshDiagnosticValidationException(message: String) : IllegalArgumentException(message)

object MeshDiagnostics {
    const val SCHEMA = "loc8.mesh-field-evidence.v1"
    const val MAX_EVENTS = 20_000

    private data class Context(
        val runId: String,
        val cohortId: String,
        val blockId: String,
        val deviceRole: String,
        val originRole: String
    )

    private var context: Context? = null
    private var active = false
    private val events = ArrayList<Map<String, Any?>>()
    private var nextEventIndex = 0
    private var nextOriginSequence = 0
    private var overflowCount = 0

    @Synchronized
    fun start(runId: String, cohortId: String, blockId: String, deviceRole: String, originRole: String) {
        if (!Regex("^[a-z0-9][a-z0-9-]{7,63}$").matches(runId)) {
            throw MeshDiagnosticValidationException("Invalid MESH-01 runId")
        }
        if (!Regex("^[a-z0-9][a-z0-9-]{0,47}$").matches(cohortId)) {
            throw MeshDiagnosticValidationException("Invalid MESH-01 cohortId")
        }
        if (!Regex("^[a-z0-9][a-z0-9-]{0,63}$").matches(blockId)) {
            throw MeshDiagnosticValidationException("Invalid MESH-01 blockId")
        }
        if (!isRole(deviceRole)) throw MeshDiagnosticValidationException("Invalid MESH-01 device role")
        if (!isRole(originRole)) throw MeshDiagnosticValidationException("Invalid MESH-01 origin role")

        if (context != null) {
            throw MeshDiagnosticValidationException("A MESH-01 snapshot is pending export/release")
        }
        context = Context(runId, cohortId, blockId, deviceRole, originRole)
        active = true
        events.clear()
        nextEventIndex = 0
        nextOriginSequence = 0
        overflowCount = 0
        append(action = "diagnostics-started")
    }

    @Synchronized
    fun stop() {
        if (!active || context == null) throw MeshDiagnosticValidationException("MESH-01 diagnostics are not active")
        append(action = "diagnostics-stopped")
        active = false
    }

    @Synchronized
    fun assertCanOriginate(sequence: Int, payload: ByteArray) {
        val ctx = context
        if (!active || ctx == null) throw MeshDiagnosticValidationException("MESH-01 diagnostics are not active")
        if (ctx.deviceRole != ctx.originRole) {
            throw MeshDiagnosticValidationException("Only the configured MESH-01 origin role may broadcast attempts")
        }
        if (sequence != nextOriginSequence) {
            throw MeshDiagnosticValidationException("MESH-01 attempts must be originated exactly once in ascending sequence")
        }
        if (!isFrozenFieldPacket(payload, ctx.originRole, sequence)) {
            throw MeshDiagnosticValidationException("MESH-01 field API accepts only the frozen zero-coordinate synthetic packet")
        }
        nextOriginSequence += 1
    }

    @Synchronized
    fun snapshot(): Map<String, Any> = mapOf(
        "events" to events.map { HashMap(it) },
        "overflowCount" to overflowCount,
        "active" to active
    )

    @Synchronized
    fun release() {
        if (active) throw MeshDiagnosticValidationException("Cannot release active MESH-01 diagnostics")
        context = null
        events.clear()
        nextEventIndex = 0
        nextOriginSequence = 0
        overflowCount = 0
    }

    @Synchronized
    fun recordLifecycle(action: String, rawLinkId: String? = null) {
        if (!active) return
        append(action = action, rawLinkId = rawLinkId)
    }

    @Synchronized
    fun recordMalformed(rawLinkId: String) {
        if (!active) return
        append(action = "malformed-drop", rawLinkId = rawLinkId, reason = "malformed")
    }

    @Synchronized
    fun recordFrame(
        action: String,
        frame: MeshFrame,
        sequence: Int? = null,
        ttlBefore: Int? = null,
        ttlAfter: Int? = null,
        rawLinkId: String? = null,
        reason: String? = null,
        onlyWhenRelayRole: Boolean = false
    ) {
        val ctx = context
        if (!active || ctx == null) return
        if (onlyWhenRelayRole && ctx.deviceRole != "B") return
        append(
            action = action,
            frame = frame,
            sequence = sequence,
            ttlBefore = ttlBefore ?: frame.ttl,
            ttlAfter = ttlAfter ?: frame.ttl,
            rawLinkId = rawLinkId,
            reason = reason
        )
    }

    private fun append(
        action: String,
        frame: MeshFrame? = null,
        sequence: Int? = null,
        ttlBefore: Int? = null,
        ttlAfter: Int? = null,
        rawLinkId: String? = null,
        reason: String? = null
    ) {
        val ctx = context ?: return
        if (events.size >= MAX_EVENTS) {
            overflowCount += 1
            return
        }
        events += linkedMapOf(
            "schema" to SCHEMA,
            "runId" to ctx.runId,
            "cohortId" to ctx.cohortId,
            "blockId" to ctx.blockId,
            "deviceRole" to ctx.deviceRole,
            "eventIndex" to nextEventIndex,
            "monotonicNs" to SystemClock.elapsedRealtimeNanos().toString(),
            "wallTimeMs" to System.currentTimeMillis(),
            "action" to action,
            "frameId" to frame?.let(::frameId),
            "sequence" to sequence,
            "originRole" to if (frame == null) null else ctx.originRole,
            "ttlBefore" to ttlBefore,
            "ttlAfter" to ttlAfter,
            "linkHandle" to rawLinkId?.let { linkHandle(ctx.runId, it) },
            "reason" to reason,
            "payloadBytes" to if (frame == null) null else MeshConstants.PAYLOAD_SIZE,
            "wireBytes" to if (frame == null) null else MeshFrameCodec.RAW_FRAME_SIZE
        )
        nextEventIndex += 1
    }

    private fun frameId(frame: MeshFrame): String {
        val identity = ArrayList<Byte>()
        identity += "LOC8-MESH-FRAME-ID-V1\u0000".toByteArray(Charsets.UTF_8).toList()
        identity += MeshConstants.PROTOCOL_VERSION
        identity += MeshConstants.MESSAGE_TYPE
        identity += MeshConstants.FLAGS
        for (shift in 56 downTo 0 step 8) {
            identity += ((frame.timestampMs ushr shift) and 0xff).toByte()
        }
        identity += frame.senderID.toList()
        identity += ((frame.payload.size ushr 8) and 0xff).toByte()
        identity += (frame.payload.size and 0xff).toByte()
        identity += frame.payload.toList()
        return sha256(identity.toByteArray())
    }

    private fun linkHandle(runId: String, rawLinkId: String): String {
        val input = runId.toByteArray(Charsets.UTF_8) + byteArrayOf(0) + rawLinkId.toByteArray(Charsets.UTF_8)
        return sha256(input).substring(0, 16)
    }

    private fun sha256(input: ByteArray): String = MessageDigest.getInstance("SHA-256")
        .digest(input)
        .joinToString("") { "%02x".format(it.toInt() and 0xff) }

    private fun isRole(value: String): Boolean = value == "A" || value == "B" || value == "C"

    private fun isFrozenFieldPacket(payload: ByteArray, role: String, sequence: Int): Boolean {
        val roleByte = when (role) {
            "A" -> 0xa1
            "B" -> 0xb1
            "C" -> 0xc1
            else -> return false
        }
        if (payload.size != MeshConstants.PAYLOAD_SIZE ||
            (payload[0].toInt() and 0xff) != 0 ||
            (payload[1].toInt() and 0xff) != 0x4c ||
            (payload[2].toInt() and 0xff) != 0x38 ||
            (payload[3].toInt() and 0xff) != 0 ||
            (payload[4].toInt() and 0xff) != roleByte ||
            (payload[19].toInt() and 0xff) != 100 ||
            (payload[24].toInt() and 0xff) != sequence
        ) return false
        for (index in 5..18) if (payload[index].toInt() != 0) return false
        return true
    }
}

File: modules/loc8-mesh/ios/MeshDiagnostics.swift
//
// MeshDiagnostics.swift
// Loc8Mesh
//
// Bounded, privacy-preserving evidence recorder for the preregistered MESH-01
// three-phone field experiment. This is observation-only: it never changes
// routing, timing, deduplication or packet bytes.
//

import CryptoKit
import Foundation

private struct MeshDiagnosticContext {
    let runID: String
    let cohortID: String
    let blockID: String
    let deviceRole: String
    let originRole: String
}

enum MeshDiagnosticValidationError: Error {
    case invalidRunID
    case invalidCohortID
    case invalidBlockID
    case invalidDeviceRole
    case invalidOriginRole
    case inactive
    case notOrigin
    case invalidOriginSequence
    case invalidFieldPacket
    case snapshotPending
}

final class MeshDiagnostics {
    static let shared = MeshDiagnostics()

    static let schema = "loc8.mesh-field-evidence.v1"
    static let maxEvents = 20_000

    private let lock = NSLock()
    private var context: MeshDiagnosticContext?
    private var active = false
    private var events: [[String: Any]] = []
    private var nextEventIndex = 0
    private var nextOriginSequence = 0
    private var overflowCount = 0

    private init() {}

    func start(runID: String, cohortID: String, blockID: String, deviceRole: String, originRole: String) throws {
        guard Self.matches(runID, pattern: "^[a-z0-9][a-z0-9-]{7,63}$") else {
            throw MeshDiagnosticValidationError.invalidRunID
        }
        guard Self.matches(cohortID, pattern: "^[a-z0-9][a-z0-9-]{0,47}$") else {
            throw MeshDiagnosticValidationError.invalidCohortID
        }
        guard Self.matches(blockID, pattern: "^[a-z0-9][a-z0-9-]{0,63}$") else {
            throw MeshDiagnosticValidationError.invalidBlockID
        }
        guard Self.isRole(deviceRole) else { throw MeshDiagnosticValidationError.invalidDeviceRole }
        guard Self.isRole(originRole) else { throw MeshDiagnosticValidationError.invalidOriginRole }

        lock.lock()
        defer { lock.unlock() }
        guard context == nil else { throw MeshDiagnosticValidationError.snapshotPending }
        context = MeshDiagnosticContext(
            runID: runID,
            cohortID: cohortID,
            blockID: blockID,
            deviceRole: deviceRole,
            originRole: originRole
        )
        active = true
        events.removeAll(keepingCapacity: true)
        nextEventIndex = 0
        nextOriginSequence = 0
        overflowCount = 0
        appendLocked(action: "diagnostics-started")
    }

    func stop() throws {
        lock.lock()
        defer { lock.unlock() }
        guard active, context != nil else { throw MeshDiagnosticValidationError.inactive }
        appendLocked(action: "diagnostics-stopped")
        active = false
    }

    func assertCanOriginate(sequence: Int, payload: Data) throws {
        lock.lock()
        defer { lock.unlock() }
        guard active, let context else { throw MeshDiagnosticValidationError.inactive }
        guard context.deviceRole == context.originRole else { throw MeshDiagnosticValidationError.notOrigin }
        guard sequence == nextOriginSequence else { throw MeshDiagnosticValidationError.invalidOriginSequence }
        guard Self.isFrozenFieldPacket(payload, role: context.originRole, sequence: sequence) else {
            throw MeshDiagnosticValidationError.invalidFieldPacket
        }
        nextOriginSequence += 1
    }

    func snapshot() -> [String: Any] {
        lock.lock()
        defer { lock.unlock() }
        return [
            "events": events,
            "overflowCount": overflowCount,
            "active": active,
        ]
    }

    func release() throws {
        lock.lock()
        defer { lock.unlock() }
        guard !active else { throw MeshDiagnosticValidationError.snapshotPending }
        context = nil
        events.removeAll(keepingCapacity: false)
        nextEventIndex = 0
        nextOriginSequence = 0
        overflowCount = 0
    }

    func recordLifecycle(action: String, rawLinkID: String? = nil) {
        lock.lock()
        defer { lock.unlock() }
        guard active else { return }
        appendLocked(action: action, rawLinkID: rawLinkID)
    }

    func recordMalformed(rawLinkID: String) {
        lock.lock()
        defer { lock.unlock() }
        guard active else { return }
        appendLocked(action: "malformed-drop", rawLinkID: rawLinkID, reason: "malformed")
    }

    func recordFrame(
        action: String,
        frame: MeshFrame,
        sequence: Int? = nil,
        ttlBefore: UInt8? = nil,
        ttlAfter: UInt8? = nil,
        rawLinkID: String? = nil,
        reason: String? = nil,
        onlyWhenRelayRole: Bool = false
    ) {
        lock.lock()
        defer { lock.unlock() }
        guard active, let context else { return }
        if onlyWhenRelayRole && context.deviceRole != "B" { return }
        appendLocked(
            action: action,
            frame: frame,
            sequence: sequence,
            ttlBefore: ttlBefore ?? frame.ttl,
            ttlAfter: ttlAfter ?? frame.ttl,
            rawLinkID: rawLinkID,
            reason: reason
        )
    }

    private func appendLocked(
        action: String,
        frame: MeshFrame? = nil,
        sequence: Int? = nil,
        ttlBefore: UInt8? = nil,
        ttlAfter: UInt8? = nil,
        rawLinkID: String? = nil,
        reason: String? = nil
    ) {
        guard let context else { return }
        guard events.count < Self.maxEvents else {
            overflowCount += 1
            return
        }

        let event: [String: Any] = [
            "schema": Self.schema,
            "runId": context.runID,
            "cohortId": context.cohortID,
            "blockId": context.blockID,
            "deviceRole": context.deviceRole,
            "eventIndex": nextEventIndex,
            "monotonicNs": String(DispatchTime.now().uptimeNanoseconds),
            "wallTimeMs": Int64(Date().timeIntervalSince1970 * 1_000),
            "action": action,
            "frameId": frame.map(Self.frameID) ?? NSNull(),
            "sequence": sequence ?? NSNull(),
            "originRole": frame == nil ? NSNull() : context.originRole,
            "ttlBefore": ttlBefore.map { Int($0) } ?? NSNull(),
            "ttlAfter": ttlAfter.map { Int($0) } ?? NSNull(),
            "linkHandle": rawLinkID.map { Self.linkHandle(runID: context.runID, rawLinkID: $0) } ?? NSNull(),
            "reason": reason ?? NSNull(),
            "payloadBytes": frame == nil ? NSNull() : MeshConstants.payloadSize,
            "wireBytes": frame == nil ? NSNull() : MeshFrameCodec.rawFrameSize,
        ]
        events.append(event)
        nextEventIndex += 1
    }

    private static func frameID(_ frame: MeshFrame) -> String {
        var identity = Data("LOC8-MESH-FRAME-ID-V1\0".utf8)
        identity.append(contentsOf: [
            MeshConstants.protocolVersion,
            MeshConstants.messageType,
            MeshConstants.flags,
        ])
        for shift in stride(from: 56, through: 0, by: -8) {
            identity.append(UInt8(truncatingIfNeeded: frame.timestampMs >> UInt64(shift)))
        }
        identity.append(frame.senderID)
        let payloadLength = UInt16(frame.payload.count)
        identity.append(UInt8(payloadLength >> 8))
        identity.append(UInt8(payloadLength & 0xff))
        identity.append(frame.payload)
        return hex(SHA256.hash(data: identity))
    }

    private static func linkHandle(runID: String, rawLinkID: String) -> String {
        var input = Data(runID.utf8)
        input.append(0)
        input.append(Data(rawLinkID.utf8))
        return String(hex(SHA256.hash(data: input)).prefix(16))
    }

    private static func hex<S: Sequence>(_ bytes: S) -> String where S.Element == UInt8 {
        bytes.map { String(format: "%02x", $0) }.joined()
    }

    private static func matches(_ value: String, pattern: String) -> Bool {
        value.range(of: pattern, options: .regularExpression) != nil
    }

    private static func isRole(_ value: String) -> Bool {
        value == "A" || value == "B" || value == "C"
    }

    private static func isFrozenFieldPacket(_ payload: Data, role: String, sequence: Int) -> Bool {
        let roleByte: UInt8
        switch role {
        case "A": roleByte = 0xa1
        case "B": roleByte = 0xb1
        case "C": roleByte = 0xc1
        default: return false
        }
        guard payload.count == MeshConstants.payloadSize,
              payload[0] == 0,
              Array(payload[1..<5]) == [0x4c, 0x38, 0x00, roleByte],
              payload[5..<19].allSatisfy({ $0 == 0 }),
              payload[19] == 100,
              payload[24] == UInt8(sequence) else {
            return false
        }
        return true
    }
}
